packages/pubsub/src/subscription.js

```diff
@@ -149,154 +144,28 @@ var PUBSUB_API_TIMEOUT = 90000;
-     *     request.
-     * @param {boolean} callback.exists - Whether the subscription exists or
-     *     not.
-     *
-     * @example
-     * subscription.exists(function(err, exists) {});
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.exists().then(function(data) {
-     *   var exists = data[0];
-     * });
-     */
-    exists: true,
-
-    /**
-     * Get a subscription if it exists.
-     *
-     * You may optionally use this to "get or create" an object by providing an
-     * object with `autoCreate` set to `true`. Any extra configuration that is
-     * normally required for the `create` method must be contained within this
-     * object as well.
-     *
-     * **`autoCreate` is only available if you accessed this object
-     * through {module:pubsub/topic#subscription}.**
-     *
-     * @param {options=} options - Configuration object.
-     * @param {boolean} options.autoCreate - Automatically create the object if
-     *     it does not exist. Default: `false`
-     *
-     * @example
-     * subscription.get(function(err, subscription, apiResponse) {
-     *   // `subscription.metadata` has been populated.
-     * });
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.get().then(function(data) {
-     *   var subscription = data[0];
-     *   var apiResponse = data[1];
-     * });
-     */
-    get: true,
-
-    /**
-     * Get the metadata for the subscription.
-     *
-     * @resource [Subscriptions: get API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/get}
-     *
-     * @param {function} callback - The callback function.
-     * @param {?error} callback.err - An error returned while making this
-     *     request.
-     * @param {?object} callback.metadata - Metadata of the subscription from
-     *     the API.
-     * @param {object} callback.apiResponse - Raw API response.
-     *
-     * @example
-     * subscription.getMetadata(function(err, metadata, apiResponse) {});
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.getMetadata().then(function(data) {
-     *   var metadata = data[0];
-     *   var apiResponse = data[1];
-     * });
-     */
-    getMetadata: {
-      protoOpts: {
-        service: 'Subscriber',
-        method: 'getSubscription'
-      },
-      reqOpts: {
-        subscription: this.name
-      }
-    }
-  };
+  this.pubsub = pubsub;
+  this.request = pubsub.request.bind(pubsub);
+  this.projectId = pubsub.projectId;
 
-  var config = {
-    parent: pubsub,
-    id: this.name,
-    methods: methods
-  };
+  this.name = Subscription.formatName_(pubsub.projectId, name);
+  this.connection = null;
+  this.encoding = options.encoding || 'utf-8';
+  this.ackDeadline = options.ackDeadline || 600;
+  this.messages_ = new Set();
```

Comment:
I agree with @jganetsk, but for a different reason. His concern about de-duplication is not actually correct since identical messages will be different objects, and thus both part of the set anyway:

```nodejs
> var s = new Set()
undefined
> s.add({})
Set { {} }
> s.add({})
Set { {}, {} }
```

That said, I still think an array would be smarter since the set loses ordering, and presumably you need the ordering, yes?

If we do not need ordering, then set is correct.